xwayland/xwayland.c

```diff
@@ -0,0 +1,227 @@
+#define _XOPEN_SOURCE 700
+#define _GNU_SOURCE
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <signal.h>
+#include <sys/socket.h>
+#include <time.h>
+#include <errno.h>
+#include <wayland-server.h>
+#include "wlr/util/log.h"
+#include "wlr/xwayland.h"
+#include "sockets.h"
+#include "xwm.h"
+
+static void safe_close(int fd) {
+	if (fd >= 0) {
+		close(fd);
+	}
+}
+
+static int unset_cloexec(int fd) {
+	if (fcntl(fd, F_SETFD, 0) != 0) {
+		wlr_log_errno(L_ERROR, "fcntl() failed on fd %d", fd);
+		return -1;
+	}
+	return 0;
+}
+
+static void exec_xwayland(struct wlr_xwayland *wlr_xwayland) {
+	if (unset_cloexec(wlr_xwayland->x_fd[0]) ||
+			unset_cloexec(wlr_xwayland->x_fd[1]) ||
+			unset_cloexec(wlr_xwayland->wm_fd[1]) ||
+			unset_cloexec(wlr_xwayland->wl_fd[1])) {
+		exit(EXIT_FAILURE);
+	}
+
+	/* Make Xwayland signal us when it's ready */
+	signal(SIGUSR1, SIG_IGN);
+
+	char *argv[11] = { 0 };
+	argv[0] = "Xwayland";
+	if (asprintf(&argv[1], ":%d", wlr_xwayland->display) < 0) {
+		wlr_log_errno(L_ERROR, "asprintf failed");
+		exit(EXIT_FAILURE);
+	}
+	argv[2] = "-rootless";
+	argv[3] = "-terminate";
+	argv[4] = "-listen";
+	if (asprintf(&argv[5], "%d", wlr_xwayland->x_fd[0]) < 0) {
+		wlr_log_errno(L_ERROR, "asprintf failed");
+		exit(EXIT_FAILURE);
+	}
+	argv[6] = "-listen";
+	if (asprintf(&argv[7], "%d", wlr_xwayland->x_fd[1]) < 0) {
+		wlr_log_errno(L_ERROR, "asprintf failed");
+		exit(EXIT_FAILURE);
+	}
+	argv[8] = "-wm";
+	if (asprintf(&argv[9], "%d", wlr_xwayland->wm_fd[1]) < 0) {
+		wlr_log_errno(L_ERROR, "asprintf failed");
+		exit(EXIT_FAILURE);
+	}
+
+	const char *xdg_runtime = getenv("XDG_RUNTIME_DIR");
+	if (!xdg_runtime) {
+		wlr_log(L_ERROR, "XDG_RUNTIME_DIR is not set");
+		exit(EXIT_FAILURE);
+	}
+
+	char *envp[3] = { 0 };
+	if (asprintf(&envp[0], "XDG_RUNTIME_DIR=%s", xdg_runtime) < 0 ||
+			asprintf(&envp[1], "WAYLAND_SOCKET=%d", wlr_xwayland->wl_fd[1]) < 0) {
+		wlr_log_errno(L_ERROR, "asprintf failed");
+		exit(EXIT_FAILURE);
+	}
+
+	wlr_log(L_INFO, "Xwayland :%d -rootless -terminate -listen %d -listen %d -wm %d",
+			wlr_xwayland->display, wlr_xwayland->x_fd[0], wlr_xwayland->x_fd[1],
+			wlr_xwayland->wm_fd[1]);
+
+	// TODO: close stdout/err depending on log level
+
+	execvpe("Xwayland", argv, envp);
+}
+
+static bool wlr_xwayland_init(struct wlr_xwayland *wlr_xwayland,
+		struct wl_display *wl_display, struct wlr_compositor *compositor);
+static void wlr_xwayland_finish(struct wlr_xwayland *wlr_xwayland);
+
+static void xwayland_destroy_event(struct wl_listener *listener, void *data) {
+	struct wl_client *client = data;
+	struct wlr_xwayland *wlr_xwayland = wl_container_of(client, wlr_xwayland, client);
+
+	/* don't call client destroy */
+	wlr_xwayland->client = NULL;
+	wlr_xwayland_finish(wlr_xwayland);
+
+	if (wlr_xwayland->server_start - time(NULL) > 5) {
+		wlr_xwayland_init(wlr_xwayland, wlr_xwayland->wl_display,
+				wlr_xwayland->compositor);
```

Comment:
Haven't tested the restart bit yet. wlc does that and it seemed like a good idea.